### PR TITLE
[Patch 2/4] Network rework: packet writing, sending and cleanups

### PR DESCRIPTION
**Pull request by nerzhul**
NetworkPacket.cpp:
* Remove some deprecated functions, we must use streaming interface
* m_data converted from u8* to std::vector<u8>
* Add an exporter to forge packet to Connection object
* implement operator << std::wstring. n
* implement operator << std::string
* dynamic resize when write packet content.
* fix string writing and performances.
* create ServerCommandFactory, used by client to get useful informations about packet processing (sending).
  * Reliability
  * Transmit channel
* Implement putRawString for some ugly char (_INIT packet), and use it.
* Many packet read and write migrated
* Implement oldForgePacket to interface writing with current connection
* fix U8/char/bool writing
* fix string writing and performances.
* add some missing functions
* Use v3s16 read instead of reading x,y,z separately
* Fix a wrong assert
* Add irr::video::SColor support into packets
* Add some missing handlers
* Add a template function to increase offset

PacketFactories:
* Create ServerCommandFactory, used by client to get useful informations about packet processing (sending).
* Create ClientCommandFactory, used by server to get useful informations about packet processing (sending).

Client.cpp:
* implement NetworkPacket ::Send interface.
* Move packet handlers to a dedicated file
* Remove Client::Send(SharedBuffer)

Server.cpp:
* implement NetworkPacket ::Send interface.
* Rewrite all packets using NetworkPacket
* Move packet handlers to a dedicated file
* Remove Server::Send(SharedBuffer)

ClientIface.cpp:
* Remove sendToAll(SharedBuffer<u8>)

Connection.hpp rework:
* Remove duplicate include
* Remove duplicate negation
* Remove a useless variable
* Improve code performance by using a m_peers_list instead of scanning m_peers map
* Remove Connection::Send(SharedBuffer)
* Fix useafterfree into NetworkPacket Sending
* Remove unused Connection::sendToAll

Test.cpp:
* Remove dead code
* Update tests to use NetworkPackets

Misc:
* add new wrappers to Send packets in client, using NetworkPacket
* Add NetworkPacket methods for Connection
* coding style fix
* dead code since changes cleanup
* Use v3s16 read instead of reading x,y,z separately in some packets
* Use different files to handle packets received by client and server
* Cleanup: Remove useless includes

### CODE REVIEW

**Comment by nerzhul** [comment]
Can somes test this commit merged with master on live servers ?

**Comment by Goofybud16** [comment]
I am guessing because of the 0.5 milestone that this breaks compatibility between vanilla 0.4.11 and this pull request?

If it doesn't, I wouldn't mind trying it if it improves performance or lowers latency.

**Comment by nerzhul** [comment]
This commit doesn't break protocol

**Comment by nerzhul** [comment]
I rebased it

**Comment by nerzhul** [comment]
Travis failed on windows because:
wget: unable to resolve host address `sfan5.pf-control.de'

**Comment by nerzhul** [comment]
Travis is now ok

**Comment by nerzhul** [comment]
@VanessaE add the PR on its running server this night

**Comment by VanessaE** [comment]
Yep, this'll go in at 0300 EST on 11 Jan.

**Comment by nerzhul** [comment]
Rebased, i fixed TOCLIENT_SPAWN_PARTICLE, which is malformed by server. Please reapply

**Comment by nerzhul** [comment]
I think we will wait until next monday (16/02) and if @VanessaE and @Zeno- doesn't report issues i will merge it.

File: src/network/packethandlers/server.cpp [hunk 1]
```diff
@@ -0,0 +1,1539 @@
...
+		m_con.DisconnectPeer(pkt->getPeerId());
+		return;
+	}
+
+	player->setPosition(position);
+	player->setSpeed(speed);
+	player->setPitch(pitch);
+	player->setYaw(yaw);
+	player->keyPressed=keyPressed;
+	player->control.up = (bool)(keyPressed & 1);
```
Comment (Zeno-) [inline src/network/packethandlers/server.cpp]: Do compilers warn if the explicit cast is not there?

**Comment by Zeno-** [comment]
+1

**Comment by nerzhul** [comment]
Merged. Commit ed04e8e9e407f0dd57fa83a9732b3a3968cb80e0

**Pull request closed without merging**